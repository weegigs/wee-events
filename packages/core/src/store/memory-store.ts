import { DateTime } from "luxon";
import { monotonicFactory } from "ulid";

import {
  AggregateId,
  DomainEvent,
  RecordedEvent,
  Revision,
} from "../types";
import { EventStore } from "./types";

const ulid = monotonicFactory();

export class MemoryStore implements EventStore {
  aggregates: Record<string, RecordedEvent[]> = {};

  clear() {
    this.aggregates = {};
  }

  async load(aggregate: AggregateId): Promise<RecordedEvent[]> {
    return this.aggregates[AggregateId.encode(aggregate)] ?? [];
  }

  async publish(aggregate: AggregateId, events: DomainEvent | DomainEvent[]): Promise<Revision> {
    const current = this.aggregates[AggregateId.encode(aggregate)] ?? [];
    const revision = current[current.length - 1]?.revision ?? Revision.Initial;

    const e = Array.isArray(events) ? events : [events];
    if (e.length == 0) {
      return revision;
    }

    const recorded: RecordedEvent[] = e.map((event) => {
      const revision = ulid()
      const timestamp = DateTime.now().toISO()

      if (timestamp == null) {
        throw new Error("unexpected invalid timestamp")
      }

      return {
        ...event,
        id: revision,
        revision,
        aggregate: aggregate,
        timestamp,
        metadata: {},
      };
    });

    const updated = current.concat(...recorded);
    this.aggregates[AggregateId.encode(aggregate)] = updated;

    return updated[updated.length - 1].revision;
  }
}
