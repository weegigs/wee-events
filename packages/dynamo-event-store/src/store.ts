import _ from "lodash";
import * as z from "zod";

import { DynamoDBClient } from "@aws-sdk/client-dynamodb";
import { DynamoDBDocumentClient, QueryCommand, TransactWriteCommand, QueryCommandOutput } from "@aws-sdk/lib-dynamodb";

import { ChangeSet } from "./changeset";

// Simple error classification functions to replace deprecated @aws-sdk/service-error-classification
const isRetryableByTrait = (error: any) => error?.retryable === true;
const isClockSkewError = (error: any) => error?.name === "ClockSkew";
const isThrottlingError = (error: any) => error?.name === "ThrottlingException" || error?.name === "ProvisionedThroughputExceededException";
const isTransientError = (error: any) => error?.name === "ServiceUnavailable" || error?.name === "InternalServerError";

import { retry } from "@weegigs/events-common";
import { Cypher, Tokenizer, no } from "@weegigs/events-cypher";

import {
  DomainEvent,
  EventStore,
  RecordedEvent,
  Revision,
  AggregateId,
  ExpectedRevisionConflictError,
  RevisionConflictError,
  Payload,
} from "@weegigs/events-core";

export namespace DynamoEventStore {
  export interface Options {
    client: () => DynamoDBDocumentClient;
    cypher: Cypher;
    tokenizer: Tokenizer;
  }
}

const defaults: DynamoEventStore.Options = {
  client: () =>
    DynamoDBDocumentClient.from(new DynamoDBClient({}), {
      marshallOptions: {
        removeUndefinedValues: true,
      },
    }),
  cypher: no.cypher,
  tokenizer: no.tokenizer,
};

type LatestRecord = {
  pk: string;
  sk: string;
  revision: string;
  timestamp: string;
};

function latestFor(changeSet: ChangeSet): LatestRecord {
  return {
    pk: changeSet.pk,
    sk: `latest-revision`,
    revision: changeSet.revision,
    timestamp: changeSet.timestamp,
  };
}

function latestCondition(changeset: ChangeSet, expectedVersion?: string) {
  if (undefined === expectedVersion) {
    return {
      ConditionExpression: "attribute_not_exists(#revision) Or #revision < :revision",
      ExpressionAttributeNames: { "#revision": "revision" },
      ExpressionAttributeValues: { ":revision": changeset.revision },
    };
  }

  if (expectedVersion === Revision.Initial) {
    return {
      ConditionExpression: "attribute_not_exists(#revision)",
      ExpressionAttributeNames: { "#revision": "revision" },
    };
  }

  return {
    ConditionExpression: "#revision = :revision",
    ExpressionAttributeNames: { "#revision": "revision" },
    ExpressionAttributeValues: { ":revision": expectedVersion },
  };
}

const isRetryableError = (error: unknown) => {
  return isRetryableByTrait(error as any) || isClockSkewError(error as any) || isThrottlingError(error as any) || isTransientError(error as any);
};

const isConditionCheckFailed = (error: unknown) =>
  (error as any)?.name === "TransactionCanceledException" &&
  (error as any).CancellationReasons?.[0]?.Code === "ConditionalCheckFailed";

export class DynamoEventStore implements EventStore {
  private readonly $client: DynamoDBDocumentClient;
  private readonly $table: string;
  private readonly $encoder: ChangeSet.EventEncoder;
  private readonly $decoder: ChangeSet.ChangeSetDecoder;

  constructor(table: string, options: Partial<DynamoEventStore.Options> = {}) {
    const { client, cypher, tokenizer } = { ...defaults, ...options };

    this.$client = client();
    this.$table = table;
    this.$encoder = ChangeSet.encoder(cypher, tokenizer);
    this.$decoder = ChangeSet.decoder(cypher);
  }

  private $read = async (
    key: string,
    decrypt: boolean,
    _after?: Revision,
    cursor: any = undefined
  ): Promise<{ events: RecordedEvent[]; next: any }> => {
    const query = new QueryCommand({
      TableName: this.$table,
      ScanIndexForward: true,
      ExclusiveStartKey: cursor,
      KeyConditionExpression: "#pk = :aggregate And begins_with(#sk, :change_set_prefix)",
      ExpressionAttributeNames: {
        "#pk": "pk",
        "#sk": "sk",
      },
      ExpressionAttributeValues: {
        ":aggregate": key,
        ":change_set_prefix": "change-set#",
      },
    });

    const { Items, LastEvaluatedKey } = await retry(
      async () => this.$client.send(query),
      (e: Error | QueryCommandOutput) => {
        return _.isError(e) ? isRetryableError(e) : false;
      },
      { limit: 5, delay: 37 }
    );

    const events: RecordedEvent[][] = await Promise.all(
      (Items ?? [])
        .map((item: any) => ChangeSet.schema.parse(item))
        .map((changeSet) => this.$decoder(changeSet, decrypt))
    );

    return { events: events.flat(), next: LastEvaluatedKey };
  };

  load = async (aggregate: AggregateId, options: EventStore.LoadOptions = {}): Promise<RecordedEvent[]> => {
    const key = AggregateId.encode(aggregate);
    const { decrypt, afterRevision: after } = { decrypt: false, ...options };

    let events: RecordedEvent[] = [];
    let cursor: any | undefined = undefined;
    do {
      const { events: page, next } = await retry(
        () => this.$read(key, decrypt, after, cursor),
        (result) => _.isError(result) && isRetryableError(result)
      );

      events = events.concat(page);
      cursor = next;
    } while (cursor);

    return events;
  };

  publish = async (
    aggregate: AggregateId,
    events: DomainEvent | DomainEvent[],
    options: EventStore.PublishOptions = {}
  ): Promise<Revision> => {
    const _events = Array.isArray(events) ? events : [events];
    if (_events.length === 0) {
      throw new Error(`expected at least one event to be published`);
    }

    const { expectedRevision } = options;

    try {
      return await retry(
        async (): Promise<Revision> => this.$publish(aggregate, _events, options),
        (e: Error | Revision) => {
          if (!_.isError(e)) {
            return false;
          }
          if (
            e.name === "TransactionCanceledException" &&
            (e as any).CancellationReasons?.[0]?.Code === "ConditionalCheckFailed"
          ) {
            return expectedRevision === undefined;
          }

          return isRetryableError(e);
        },
        { limit: 5, delay: 37 }
      );
    } catch (e: unknown) {
      if (_.isError(e) && isConditionCheckFailed(e)) {
        if (expectedRevision !== undefined) {
          throw new ExpectedRevisionConflictError(expectedRevision);
        }

        throw new RevisionConflictError();
      }

      throw e;
    }
  };

  private async $publish(
    aggregate: AggregateId,
    events: DomainEvent[],
    options: EventStore.PublishOptions = {}
  ): Promise<Revision> {
    const { encrypt, expectedRevision } = options;

    const recorded: RecordedEvent[] = await Promise.all(
      events
        .map((event) => {
          // Validate that the event has the proper DomainEvent structure
          const domainEventSchema = z.object({
            type: z.string().min(1),
            data: Payload.schema
          });
          return domainEventSchema.parse(event);
        })
        .map((event) =>
          this.$encoder(aggregate, event, _.pick(options, "correlationId", "causationId"), encrypt ?? false)
        )
    );

    const changeSet = await ChangeSet.create(aggregate, recorded);
    return this.$write(changeSet, expectedRevision);
  }

  private async $write(changeSet: ChangeSet, expectedRevision?: string): Promise<Revision> {
    const TableName = this.$table;

    const latest = latestFor(changeSet);
    const latestCheck = latestCondition(changeSet, expectedRevision);

    const write = new TransactWriteCommand({
      TransactItems: [
        {
          Put: {
            TableName,
            Item: latest,
            ...latestCheck,
          },
        },
        {
          Put: {
            TableName,
            Item: changeSet,
          },
        },
      ],
    });

    await this.$client.send(write);

    return changeSet.revision;
  }
}
