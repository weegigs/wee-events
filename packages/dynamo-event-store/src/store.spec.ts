import _ from "lodash";

import { range } from "@weegigs/events-common";
import { Cypher, Tokenizer } from "@weegigs/events-cypher";

import { CreateTableCommand, DynamoDBClient, DescribeTableCommand } from "@aws-sdk/client-dynamodb";
import { DynamoDBDocumentClient } from "@aws-sdk/lib-dynamodb";
import { DomainEvent, Payload, Revision, AggregateId, Errors } from "@weegigs/events-core";
import { GenericContainer, StartedTestContainer } from "testcontainers";

import { DynamoEventStore } from "./store";

// it takes time to spin the containers up
jest.setTimeout(100000);

const sampleEvent: DomainEvent = {
  type: "test:test-event",
  data: {
    string: "a string",
    number: 123,
    boolean: true,
    nested: {
      string: "a nested string",
      number: 72600,
    },
    array: ["one"],
  },
};

let id: string;
function testStream(suffix = ""): AggregateId {
  return { type: "test", key: `${id}${suffix}` };
}

beforeEach(() => {
  id = Date.now().toString() + Math.random().toString(36).substring(2);
});

const tokenizer: Tokenizer = Tokenizer.create(async (path, value) => {
  return `${path}:${value}`;
});

const cypher: Cypher = {
  async decrypt(_streamId: AggregateId, buffer: Buffer): Promise<Payload> {
    const decoded = buffer.toString("utf-8");
    return JSON.parse(decoded.split("").reverse().join(""));
  },
  async encrypt(_streamId: AggregateId, payload: Payload): Promise<Buffer> {
    const encrypted = JSON.stringify(payload).split("").reverse().join("");
    return Buffer.from(encrypted, "utf-8");
  },
};

const TableName = "test-events";

describe("dynamo store", () => {
  let dbc: StartedTestContainer;
  let client: DynamoDBDocumentClient;

  beforeAll(async () => {
    dbc = await new GenericContainer("amazon/dynamodb-local")
      .withExposedPorts(8000)
      .withCommand(["-jar", "DynamoDBLocal.jar", "-sharedDb", "-inMemory"])
      .start();
    const dynamo = new DynamoDBClient({
      endpoint: `http://${dbc.getHost()}:${dbc.getMappedPort(8000)}`,
      credentials: { 
        accessKeyId: "test", 
        secretAccessKey: "test" 
      },
      region: "us-east-1"
    });

    const create = new CreateTableCommand({
      TableName,
      KeySchema: [
        { AttributeName: "pk", KeyType: "HASH" }, //Partition key
        { AttributeName: "sk", KeyType: "RANGE" }, //Sort key
      ],
      AttributeDefinitions: [
        { AttributeName: "pk", AttributeType: "S" },
        { AttributeName: "sk", AttributeType: "S" },
      ],
      ProvisionedThroughput: {
        ReadCapacityUnits: 10,
        WriteCapacityUnits: 10,
      },
    });

    await dynamo.send(create);

    let ready = false;
    do {
      const describe = new DescribeTableCommand({ TableName });
      const { Table } = await dynamo.send(describe);

      ready = Table?.TableStatus === "ACTIVE";
    } while (!ready);

    client = DynamoDBDocumentClient.from(dynamo, {
      marshallOptions: {
        removeUndefinedValues: true,
      },
    });
  });

  afterAll(async () => {
    await dbc.stop();
  });

  const createStore = (options: Partial<Omit<DynamoEventStore.Options, "client">> = {}) =>
    new DynamoEventStore(TableName, { ...options, client: () => client });

  describe("event publishing", () => {
    it("publishes basic events", async () => {
      const store = createStore();

      const revision = await store.publish(testStream(), sampleEvent);

      expect(revision).toBeDefined();
      expect(revision).toHaveLength(26);
    });

    it("rejects events with an incorrect expected revision", async () => {
      const store = createStore();

      try {
        await store.publish(testStream(), sampleEvent, {
          expectedRevision: "bob",
        });
        throw new Error("expected publish to fail");
      } catch (e: unknown) {
        expect((e as Error).message).toEqual("revision did not match the expected revision of bob");
        expect((e as any).name).toEqual("expected-revision-conflict");
        expect(Errors.isExpectedRevisionConflict(e as any)).toBeTruthy();
      }
    });

    it("rejects events with when expecting an empty aggregate", async () => {
      const store = createStore();

      await store.publish(testStream(), sampleEvent);
      try {
        await store.publish(testStream(), sampleEvent, {
          expectedRevision: Revision.Initial,
        });
        throw new Error("expected publish to fail");
      } catch (e: unknown) {
        expect((e as Error).message).toEqual("revision did not match the expected revision of 00000000000000000000000000");
        expect((e as any).name).toEqual("expected-revision-conflict");
        expect(Errors.isExpectedRevisionConflict(e as any)).toBeTruthy();
      }
    });

    it("publishes events with an correct expected revision", async () => {
      const store = createStore();

      const first = await store.publish(testStream(), sampleEvent, {});
      const second = await store.publish(testStream(), sampleEvent, {
        expectedRevision: first,
      });

      expect(second).toBeDefined();
      expect(second).toHaveLength(26);
    });

    it("rejects events with invalid data", async () => {
      const store = createStore();
      try {
        await store.publish(testStream(), { type: "invalid-data", data: { value: null } } as any);
        throw new Error("expected publish to fail");
      } catch (e: unknown) {
        expect(e).toMatchSnapshot();
      }
    });

    it("rejects events with invalid type", async () => {
      const store = createStore();
      try {
        await store.publish(testStream(), { type: "", data: {} });
        throw new Error("expected publish to fail");
      } catch (e: unknown) {
        expect(e).toMatchSnapshot();
      }
    });
  });

  describe("aggregate loading", () => {
    it("loads streams with no events", async () => {
      const store = createStore();
      const events = await store.load(testStream("-new"));

      expect(events).toHaveLength(0);
    });

    it("loads streams with one event", async () => {
      const store = createStore();
      await store.publish(testStream("-existing"), sampleEvent);

      const events = await store.load(testStream("-existing"));

      expect(events).toBeDefined();
      expect(events).toHaveLength(1);
    });

    it("loads streams with multiple events", async () => {
      const store = createStore();
      for (let i = 0; i < 5; i++) {
        await store.publish(testStream("-multiple"), sampleEvent);
      }

      const events = await store.load(testStream("-multiple"));

      expect(events).toHaveLength(5);
      expect(_.sortBy(events, (e) => e.timestamp)).toEqual(events);
    });

    it("loads streams with events published in a batch", async () => {
      const store = createStore();

      for (let i = 0; i < 5; i++) {
        await store.publish(
          testStream("-batch"),
          Array(...range(5)).map((_) => sampleEvent)
        );
      }

      const events = await store.load(testStream("-batch"));

      expect(events).toHaveLength(25);
      expect(_.sortBy(events, (e) => e.timestamp)).toEqual(events);
    });
  });

  describe("correlation", () => {
    it("should connect correlation ids to events", async () => {
      const store = createStore();
      await store.publish(testStream("-correlate"), sampleEvent, { correlationId: "bob" });

      const events = await store.load(testStream("-correlate"));

      expect(events).toBeDefined();
      expect(events).toHaveLength(1);
      expect(events[0].metadata.correlationId).toEqual("bob");
    });

    it("should connect causation ids to events", async () => {
      const store = createStore();
      await store.publish(testStream("-causation"), sampleEvent, { causationId: "123" });

      const events = await store.load(testStream("-causation"));

      expect(events).toBeDefined();
      expect(events).toHaveLength(1);
      expect(events[0].metadata.causationId).toEqual("123");
    });

    it("should connect both correlation & causation ids to events", async () => {
      const store = createStore();
      await store.publish(testStream("-correlation-causation"), sampleEvent, {
        correlationId: "bob",
        causationId: "123",
      });

      const events = await store.load(testStream("-correlation-causation"));

      expect(events).toBeDefined();
      expect(events).toHaveLength(1);
      expect(events[0].metadata.correlationId).toEqual("bob");
      expect(events[0].metadata.causationId).toEqual("123");
    });
  });

  describe("encryption", () => {
    it("should tokenize the content of encrypted events", async () => {
      const store = createStore({ tokenizer, cypher });
      await store.publish(testStream("-tokenize"), sampleEvent, { encrypt: true });

      const events = await store.load(testStream("-tokenize"));
      expect(events).toHaveLength(1);
      expect(events[0].data).not.toEqual(sampleEvent.data);
      expect(events[0].data).toMatchSnapshot();
    });

    it("should allow you to decrypt events", async () => {
      const store = createStore({ tokenizer, cypher });
      await store.publish(testStream("-decrypt"), sampleEvent, { encrypt: true });

      const events = await store.load(testStream("-decrypt"), { decrypt: true });
      expect(events).toHaveLength(1);
      expect(events[0].data).toEqual(sampleEvent.data);
    });

    it("should process streams with both encrypted and unencrypted events", async () => {
      const store = createStore({ tokenizer, cypher });
      await store.publish(testStream("-maybe-decrypt"), sampleEvent);
      await store.publish(testStream("-maybe-decrypt"), sampleEvent, { encrypt: true });

      const events = await store.load(testStream("-maybe-decrypt"), { decrypt: true });
      expect(events).toHaveLength(2);
      expect(events[0].data).toEqual(sampleEvent.data);
      expect(events[1].data).toEqual(sampleEvent.data);
    });
  });
});
