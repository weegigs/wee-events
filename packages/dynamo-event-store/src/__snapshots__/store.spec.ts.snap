// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`dynamo store > encryption > should tokenize the content of encrypted events 1`] = `
{
  "array": [
    "array.0:one",
  ],
  "boolean": true,
  "nested": {
    "number": -105807839,
    "string": "nested.string:a nested string",
  },
  "number": -393008511,
  "string": "string:a string",
}
`;

exports[`dynamo store > event publishing > rejects events with invalid data 1`] = `
[ZodError: [
  {
    "code": "invalid_union",
    "unionErrors": [
      {
        "issues": [
          {
            "code": "invalid_union",
            "unionErrors": [
              {
                "issues": [
                  {
                    "code": "invalid_type",
                    "expected": "string",
                    "received": "null",
                    "path": [
                      "data",
                      "value"
                    ],
                    "message": "Expected string, received null"
                  }
                ],
                "name": "ZodError"
              },
              {
                "issues": [
                  {
                    "code": "invalid_type",
                    "expected": "number",
                    "received": "null",
                    "path": [
                      "data",
                      "value"
                    ],
                    "message": "Expected number, received null"
                  }
                ],
                "name": "ZodError"
              },
              {
                "issues": [
                  {
                    "code": "invalid_type",
                    "expected": "boolean",
                    "received": "null",
                    "path": [
                      "data",
                      "value"
                    ],
                    "message": "Expected boolean, received null"
                  }
                ],
                "name": "ZodError"
              }
            ],
            "path": [
              "data",
              "value"
            ],
            "message": "Invalid input"
          }
        ],
        "name": "ZodError"
      },
      {
        "issues": [
          {
            "code": "invalid_union",
            "unionErrors": [
              {
                "issues": [
                  {
                    "code": "invalid_type",
                    "expected": "array",
                    "received": "null",
                    "path": [
                      "data",
                      "value"
                    ],
                    "message": "Expected array, received null"
                  }
                ],
                "name": "ZodError"
              },
              {
                "issues": [
                  {
                    "code": "invalid_type",
                    "expected": "array",
                    "received": "null",
                    "path": [
                      "data",
                      "value"
                    ],
                    "message": "Expected array, received null"
                  }
                ],
                "name": "ZodError"
              },
              {
                "issues": [
                  {
                    "code": "invalid_type",
                    "expected": "array",
                    "received": "null",
                    "path": [
                      "data",
                      "value"
                    ],
                    "message": "Expected array, received null"
                  }
                ],
                "name": "ZodError"
              }
            ],
            "path": [
              "data",
              "value"
            ],
            "message": "Invalid input"
          }
        ],
        "name": "ZodError"
      },
      {
        "issues": [
          {
            "code": "invalid_type",
            "expected": "object",
            "received": "null",
            "path": [
              "data",
              "value"
            ],
            "message": "Expected object, received null"
          }
        ],
        "name": "ZodError"
      }
    ],
    "path": [
      "data",
      "value"
    ],
    "message": "Invalid input"
  }
]]
`;

exports[`dynamo store > event publishing > rejects events with invalid type 1`] = `
[ZodError: [
  {
    "code": "too_small",
    "minimum": 1,
    "type": "string",
    "inclusive": true,
    "exact": false,
    "message": "String must contain at least 1 character(s)",
    "path": [
      "type"
    ]
  }
]]
`;
